### PR TITLE
Add register_dir{,s}_for_cleanup

### LINKED ISSUE #8018: Feature request: register_dir_for_cleanup
Currently the filedropper mixin makes `register_files_for_cleanup` available to clean dropped files after a session is created.

Some plugins also need to remove a created folder so a `register_dir_for_cleanup` method would be really handy.

Two plugins I currently know that can use this because the plugin upload creates a folder with the plugin name (in both cases) and the empty folder stays on the machine after running the module:
* https://github.com/rapid7/metasploit-framework/blob/master/modules/exploits/unix/webapp/piwik_superuser_plugin_upload.rb
* https://github.com/rapid7/metasploit-framework/blob/master/modules/exploits/unix/webapp/wp_admin_shell_upload.rb

### PR DESCRIPTION
**Pull request by wvu**
I lost the original patch and had to rewrite it. I have no idea if this works...

Fixes #8018?

### CODE REVIEW

**Comment by bcoles** [comment]
This didn't work as described.

Tested on Windows 7 SP1 with PHP meterpreter payload.

On the bright side, applying the changes in this PR caused `register_dir_for_cleanup` to operate in the same fashion as `register_file_for_cleanup`.

Unfortunately, it did this by rendering `register_file_for_cleanup` useless, preventing it from cleaning up files.

File: lib/msf/core/exploit/file_dropper.rb [hunk 1]
```diff
@@ -134,7 +166,7 @@ def file_dropper_file_exist?(session, path)
   #
   # @param [String] file The file to delete
   # @return [Boolean] True if the delete command has been executed in the remote machine, otherwise false.
-  def file_dropper_delete(session, file)
+  def file_dropper_delete_file(session, file)
     win_file = file_dropper_win_file(file)
```
Comment (bcoles) [inline lib/msf/core/exploit/file_dropper.rb]: `file_dropper_win_file` is no longer defined. It has been renamed to `file_dropper_win_path`

Comment (wvu-r7) [reply lib/msf/core/exploit/file_dropper.rb]: Oops, I missed one.

**Comment by bcoles** [comment]
Tested and working successfully with PHP Meterpreter on WIndows 7 SP1.

(after changing the one instance of `file_dropper_win_file` to `file_dropper_win_path`)

**Comment by wvu-r7** [comment]
Amazing. I didn't even test it yet. Try Linux?

**Comment by bcoles** [comment]
Trying PHP Meterpreter on Debian 8.20 now.

**Comment by bcoles** [comment]
Tested and working successfully with PHP Meterpreter on Debian 8.2.0.

**Comment by bcoles** [comment]
Failed with Powershell Meterpreter on Windows 7 SP1

```
[-] 172.16.191.166:2810 - Exploit failed: undefined method `+' for nil:NilClass
```

**Comment by wvu-r7** [comment]
Hmm, got a stack trace? I think I know what that one is.

```
127     (@dropped_files + @dropped_dirs).each do |p|
```

**Comment by bcoles** [comment]
Summarizing our discussion from IRC here for context:

> In hindsight, the issue was not related to Powershell. My initial tests were with done with both `register_file_for_cleanup` and `register_dir_for_cleanup`, where as the Powershell test was done with only the latter.

New tests, with Powershell Meterpreter on Windows 7 SP1:

### `register_file_for_cleanup` only : deleted the file :

`register_file_for_cleanup 'C:\test\asdf.txt'`

```
[*] Started reverse TCP handler on 172.16.191.238:4444 
[*] 172.16.191.166:2810 - Sending payload (6111 bytes)...
[*] Sending stage (179779 bytes) to 172.16.191.166
[*] Meterpreter session 1 opened (172.16.191.238:4444 -> 172.16.191.166:51741) at 2018-01-06 05:12:25 -0500
[!] 172.16.191.166:2810 - This exploit may require manual cleanup of 'C:\test\asdf.txt' on the target

meterpreter > 
[+] 172.16.191.166:2810 - Deleted C:\test\asdf.txt
```

### `register_dir_for_cleanup` only : deleted the directory :

`register_dir_for_cleanup 'C:\test'`

```
[*] Started reverse TCP handler on 172.16.191.238:4444 
[*] 172.16.191.166:2810 - Sending payload (6103 bytes)...
[*] Sending stage (179779 bytes) to 172.16.191.166
[*] Meterpreter session 2 opened (172.16.191.238:4444 -> 172.16.191.166:51770) at 2018-01-06 05:16:43 -0500
[!] 172.16.191.166:2810 - This exploit may require manual cleanup of 'C:\test' on the target

meterpreter > 
[+] 172.16.191.166:2810 - Deleted C:\test
```

### `register_file_for_cleanup` then `register_dir_for_cleanup` : deleted both

```
register_file_for_cleanup 'C:\asdf.txt'
register_dir_for_cleanup 'C:\test'
```

```
[*] Started reverse TCP handler on 172.16.191.238:4444 
[*] 172.16.191.166:2810 - Sending payload (6111 bytes)...
[*] Sending stage (179779 bytes) to 172.16.191.166
[*] Meterpreter session 3 opened (172.16.191.238:4444 -> 172.16.191.166:51797) at 2018-01-06 05:20:25 -0500
[!] 172.16.191.166:2810 - This exploit may require manual cleanup of 'C:\asdf.txt' on the target
[!] 172.16.191.166:2810 - This exploit may require manual cleanup of 'C:\test' on the target

meterpreter > 
[+] 172.16.191.166:2810 - Deleted C:\asdf.txt
[+] 172.16.191.166:2810 - Deleted C:\test
```

### `register_dir_for_cleanup` then `register_file_for_cleanup` : deleted both

```
register_dir_for_cleanup 'C:\test'
register_file_for_cleanup 'C:\asdf.txt'
```

```
[*] Started reverse TCP handler on 172.16.191.238:4444 
[*] 172.16.191.166:2810 - Sending payload (6103 bytes)...
[*] Sending stage (179779 bytes) to 172.16.191.166
[*] Meterpreter session 5 opened (172.16.191.238:4444 -> 172.16.191.166:51809) at 2018-01-06 05:21:34 -0500
[!] 172.16.191.166:2810 - This exploit may require manual cleanup of 'C:\asdf.txt' on the target
[!] 172.16.191.166:2810 - This exploit may require manual cleanup of 'C:\test' on the target

meterpreter > 
[+] 172.16.191.166:2810 - Deleted C:\asdf.txt
[+] 172.16.191.166:2810 - Deleted C:\test
```

**Review by bcoles — APPROVED** [review]
I approve these changes, on the condition that everyone yell at @wvu-r7 if it breaks things.

**Comment by wvu-r7** [comment]
Someone land this so I can fix more stuff when it breaks. :-)

**Comment by wvu-r7** [comment]
# Release Notes

This adds the `register_dirs_for_cleanup` (alias `register_dir_for_cleanup`) method to `FileDropper`, giving module authors the ability to clean up dropped directories after exploitation.

**Comment by busterb** [comment]
A good time to add some PRs to convert modules while the iron is hot?

**Comment by wvu-r7** [comment]
I know I'll be converting mine. Did you have any of your own, @bcoles && @FireFart?

**Comment by bcoles** [comment]
Just #8539 pending merge. Already updated.

**Comment by FireFart** [comment]
Awesome thx! I will go through my modules and add it

**Pull request merged**